Revamp sending-a-message-to-a-service tutorial (#8057)
ballerina-platform/ballerina-dev-website: swan-lake/integration-tutorials/revamped-sending-a-message-to-a-service.md

---

## Overview

In this tutorial, you will develop a service that allows a user to retrieve a list of doctors based on the doctor's specialization (category). The information about the doctors is retrieved from a separate microservice. 

To implement this use case, you will develop a REST service with a single resource using Visual Studio Code with the Ballerina Swan Lake extension, and then, run the service. The resource will receive the user request, retrieve details from the backend service, and respond to the user request with the relevant doctor details.

### Concepts covered

- REST API
- HTTP client

## Develop the application

### Step 1: Set up the workspace

Install [Ballerina Swan Lake](https://ballerina.io/downloads/) and the [Ballerina Swan Lake extension](https://marketplace.visualstudio.com/items?itemName=wso2.ballerina) on VS Code.

### Step 2: Develop the service

Follow the instructions given in this section to develop the service.

1. Create a new Ballerina project using the `bal` command and open it in VS Code.

    ```bash
    $ bal new sending-a-message-to-a-service

>     ```

Reviewer: ````suggestion
    ```

````

> 2. Remove the example content and open Ballerina diagram view with [Ballerina HTTP API Designer](https://wso2.com/ballerina/vscode/docs/design-the-services/http-api-designer) in VS Code.

Reviewer: ```suggestion
2. Remove the generated content in the `main.bal` file and open the [Ballerina HTTP API Designer](https://wso2.com/ballerina/vscode/docs/design-the-services/http-api-designer) in VS Code.
```

    ![Open diagram view](/learn/images/tutorial_sending_a_message_to_a_service_open_diagram_view.gif)

3. Define two [configurable variables](https://ballerina.io/learn/by-example/#configurability) for the port on which the listener should listen and the URL of the backend service.

    ![Define configurable variables](/learn/images/tutorial_sending_a_message_to_a_service_define_configurable_variables.gif)

>     The source code we generated is as follows.

Reviewer: ```suggestion
    The corresponding source code that is generated is as follows.
```

    ```ballerina
    configurable int port = 8290;
    configurable string healthcareBackend = "http://localhost:9090/healthcare";
    ```

4. Define a record corresponding to the payload from the backend service.

    ```ballerina
    type Doctor record {|
        string name;
        string hospital;
        string category;
        string availability;
        decimal fee;
    |};
    ```
    

>     > **Note:** You can use the "Paste JSON as record" VS Code command to generate the record if you have the JSON payload.

Reviewer: We discussed using sample JSON payload as the default, right?

So 

4. Generate the record corresponding to the payload from the backend service using the "Paste JSON as record" VS Code command by providing a sample of the expected JSON payload.

At the end we can say, if there is not sample JSON they can manually define the record. Basically, prioritize using the command.

    > ![Paste JSON as record](/learn/images/tutorial_sending_a_message_to_a_service_paste_json_as_record.gif)
    > 
    > 1. Copy the JSON with `ctrl+c` or `cmd+c`.
    > 2. Open Command pallette using `Ctrl+Shift+P` or `Cmd+Shift+P`.
    > 3. Search for `Ballerina: Paste JSON as record` and select it or hit `Enter`.

    The payload will be an array of JSON objects in which, the structure of each JSON object matches this record.

5. Define the [HTTP service (REST API)](https://ballerina.io/learn/by-example/#rest-service) that has the resource that accepts user requests, retrieves relevant details from the backend service, and responds to the request.

    - Use `/healthcare` as the service path (or the context) of the service, which is attached to the listener listening on port `port`.

        ![Define the service](/learn/images/tutorial_sending_a_message_to_a_service_define_a_service.gif)

    - Define an HTTP resource that allows the `GET` operation on resource path `/querydoctor` and accepts the `category` (corresponding to the specialization) as a path parameter.

        ![Define the resource](/learn/images/tutorial_sending_a_message_to_a_service_define_a_resource.gif)

>     Generated service will be as follows.

Reviewer: ```suggestion
    The generated service will be as follows.
```

>     ```ballerina

Reviewer: Shouldn't this be indented?

    service /healthcare on new http:Listener(port) {
        resource function get querydoctor/[string category]() 
                returns Doctor[]|http:NotFound|http:InternalServerError {
            
        }
    }
    ```

> 6. Before implementing the logic, define an [`http:Client` client](https://ballerina.io/learn/by-example/#http-client) to send requests to the backend service.

Reviewer: ```suggestion
6. Define an [`http:Client` client](https://ballerina.io/learn/by-example/#http-client) to send requests to the backend service.
```

>     ```ballerina

Reviewer: Why are we not using the low code approach here?

Author: Felt odd to write a long-expression and use references (`halthcareBackend`), since we didn't get any suggestions.

![Peek 2023-10-18 10-20](https://github.com/ballerina-platform/ballerina-dev-website/assets/25485997/62a17e71-4e3d-47bf-931c-d4b30059d11a)

Reviewer: Not like this. As a connector.

+ Component -> Module level variables - Connector -> Search for http and add

Author: Ack

    final http:Client queryDoctorEP = check new (healthcareBackend);
    ```

    > **Note:** The argument to the `new` expression is the URL for the backend service. 
    > 
    > Later, the client will be initialized using a seperate function to aid with testing.
    > 
    >   ```ballerina
    >   final http:Client queryDoctorEP = check initializeHttpClient();
    >
    >   function initializeHttpClient(string url) returns http:Client|error => new (healthcareBackend);

>     >   ```

Reviewer: We said we'll move this to the testing section, right?
